akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -489,6 +486,13 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
         None // don't change state
       }
     }
+    def onTrailingHeaders(headers: ParsedHeadersFrame): Unit = {
+      if (headers.endStream) {
+        onDataFrame(DataFrame(headers.streamId, endStream = true, ByteString.empty)) // simulate end stream by empty dataframe
+      } else pushGOAWAY(Http2Protocol.ErrorCode.PROTOCOL_ERROR, "Got unexpected mid-stream HEADERS frame")
+
+      trailer.success(headers)
```

Comment:
What happens in the case where the last frame is a DATA frame with `endStream = true`? Shouldn't `trailer` be completed in that case as well?

Reply:
(Looking through IncomingStreamState, there seem to be a few loopholes, where we don't go through the state machine which should leave the stream state in a non-closed state after the final response data frame has been sent out)